### PR TITLE
Rework synchronization: first spin, then yield, and finally block.

### LINKED ISSUE #838: adding constant to an array: performance regression compared to CUDAdrv
Adding constant to an array runs almost 20% slower in CUDA.jl compared to CUDAdrv.jl
```
function add_const!(y, x, p)
    y .= x .+ p 
end

#test setup
Random.seed!(1)
m, n = 3072, 1536    # 256 multiplier
x1 = cu(randn(Float32, (m, n)) .+ Float32(0.5))
y1 = similar(x1)

add_const!(y1, x1, 1f0) #warmup
print("add!                       "); 
@btime begin add_const!($y1, $x1, 1f0); synchronize() end
```
with CUDAdrv.jl
`add!                         121.000 μs (60 allocations: 2.33 KiB)`
with CUDA.jl
`add!                         141.500 μs (219 allocations: 3.52 KiB)`

```
julia> CUDA.versioninfo()
CUDA toolkit 11.2.2, artifact installation
CUDA driver 11.3.0

Libraries:        
- CUBLAS: 11.4.1
- CURAND: 10.2.3
- CUFFT: 10.4.1
- CUSOLVER: 11.1.0
- CUSPARSE: 11.4.1
- CUPTI: 14.0.0
- NVML: missing
- CUDNN: 8.10.0 (for CUDA 11.2.0)
- CUTENSOR: 1.2.2 (for CUDA 11.1.0)

Toolchain:
- Julia: 1.6.0
- LLVM: 11.0.1
- PTX ISA support: 3.2, 4.0, 4.1, 4.2, 4.3, 5.0, 6.0, 6.1, 6.3, 6.4, 6.5, 7.0
- Device support: sm_35, sm_37, sm_50, sm_52, sm_53, sm_60, sm_61, sm_62, sm_70, sm_72, sm_75, sm_80

1 device:
  0: NVIDIA GeForce RTX 2070 (sm_75, 6.847 GiB / 8.000 GiB available
```
```
julia> CUDA.versioninfo()
CUDA toolkit 11.2.2, artifact installation
CUDA driver 11.3.0

Libraries:        
- CUBLAS: 11.4.1
- CURAND: 10.2.3
- CUFFT: 10.4.1
- CUSOLVER: 11.1.0
- CUSPARSE: 11.4.1
- CUPTI: 14.0.0
- NVML: missing
- CUDNN: 8.10.0 (for CUDA 11.2.0)
- CUTENSOR: 1.2.2 (for CUDA 11.1.0)

Toolchain:
- Julia: 1.6.0
- LLVM: 11.0.1
- PTX ISA support: 3.2, 4.0, 4.1, 4.2, 4.3, 5.0, 6.0, 6.1, 6.3, 6.4, 6.5, 7.0
- Device support: sm_35, sm_37, sm_50, sm_52, sm_53, sm_60, sm_61, sm_62, sm_70, sm_72, sm_75, sm_80

1 device:
  0: NVIDIA GeForce RTX 2070 (sm_75, 6.847 GiB / 8.000 GiB available](url)
```

full code (setup + run) is [here](https://1drv.ms/u/s!AkkVaBzPZViciaY-jLRlB4_gnE5f0Q?e=VHsVL8) you run it like this
```
C:\Bin\Julia-1.3.1\bin\julia.exe add_kernel_1.3.1_setup.jl
julia add_kernel_1.6.0_setup.jl

C:\Bin\Julia-1.3.1\bin\julia.exe add_kernel_1.3.1.jl
julia add_kernel_1.6.0.jl
```

### LINKED ISSUE #839: CUDA.abs() on vector input: performance regression compared to CUDAdrv
Below code runs about 10% slower than using `CUDAnative.abs` and `CUDAdrv` in case the input is a vector (there is only about 1% performance difference in case of matrix input)
```
const threads = 256
function kernel_abs(n, x, y)
    i = (blockIdx().x-1) * blockDim().x + threadIdx().x
    if i <= n
        @inbounds y[i] = CUDA.abs(x[i])
    end
    return
end

function abs!(y, x)
    len = length(x)
    blocks = (len + threads - 1) ÷ threads
    @cuda blocks = blocks, 1 threads = threads kernel_abs(len, x, y)
end

#test setup
Random.seed!(1)
x1 = cu(randn(Float32, (1, 1536)) .+ Float32(0.5))
y1 = similar(x1)

abs!(y1, x1) #warmup
print("add!                       "); 
@btime begin abs!($y1, $x1); synchronize() end
```

with CUDAdrv.jl
`abs!                         18.000 μs (29 allocations: 976 bytes)`
with CUDA.jl
`abs!                         19.900 μs (12 allocations: 288 bytes)`

```
"CUDA"         => v"3.0.2"
"BenchmarkTools"  => v"0.6.0"
```
```
julia> CUDA.versioninfo()
CUDA toolkit 11.2.2, artifact installation
CUDA driver 11.3.0

Libraries:        
- CUBLAS: 11.4.1
- CURAND: 10.2.3
- CUFFT: 10.4.1
- CUSOLVER: 11.1.0
- CUSPARSE: 11.4.1
- CUPTI: 14.0.0
- NVML: missing
- CUDNN: 8.10.0 (for CUDA 11.2.0)
- CUTENSOR: 1.2.2 (for CUDA 11.1.0)

Toolchain:
- Julia: 1.6.0
- LLVM: 11.0.1
- PTX ISA support: 3.2, 4.0, 4.1, 4.2, 4.3, 5.0, 6.0, 6.1, 6.3, 6.4, 6.5, 7.0
- Device support: sm_35, sm_37, sm_50, sm_52, sm_53, sm_60, sm_61, sm_62, sm_70, sm_72, sm_75, sm_80

1 device:
  0: NVIDIA GeForce RTX 2070 (sm_75, 6.847 GiB / 8.000 GiB available
```
```
julia> CUDA.versioninfo()
CUDA toolkit 11.2.2, artifact installation
CUDA driver 11.3.0

Libraries:        
- CUBLAS: 11.4.1
- CURAND: 10.2.3
- CUFFT: 10.4.1
- CUSOLVER: 11.1.0
- CUSPARSE: 11.4.1
- CUPTI: 14.0.0
- NVML: missing
- CUDNN: 8.10.0 (for CUDA 11.2.0)
- CUTENSOR: 1.2.2 (for CUDA 11.1.0)

Toolchain:
- Julia: 1.6.0
- LLVM: 11.0.1
- PTX ISA support: 3.2, 4.0, 4.1, 4.2, 4.3, 5.0, 6.0, 6.1, 6.3, 6.4, 6.5, 7.0
- Device support: sm_35, sm_37, sm_50, sm_52, sm_53, sm_60, sm_61, sm_62, sm_70, sm_72, sm_75, sm_80

1 device:
  0: NVIDIA GeForce RTX 2070 (sm_75, 6.847 GiB / 8.000 GiB available](url)
```

full code (setup + run) is [here](https://1drv.ms/u/s!AkkVaBzPZViciaZHsZLXzcdnQ-_YlQ?e=B6nFZ6) you run it like this
```
C:\Bin\Julia-1.3.1\bin\julia.exe run_kernel_1.3.1_setup.jl
julia run_kernel_1.6.0_setup.jl

C:\Bin\Julia-1.3.1\bin\julia.exe run_kernel_1.3.1.jl
julia run_kernel_1.6.0.jl
```

Julia 1.3.1 setup
```
  [6e4b80f9] BenchmarkTools v0.5.0
  [c5f51814] CUDAdrv v5.0.1
  [be33ccc6] CUDAnative v2.7.0
  [3a865a2d] CuArrays v1.6.0
```

### LINKED ISSUE #893: CUDA.@sync seems to be using a lot of CPU while waiting
The following code seems to be loading a CPU thread fully. (1/16th on 16th thread CPU and if I run 6 processes like this it is using 50+% on the same cpu). The expectation is that while processing is happening in GPU there will be relatively low CPU load.
```
Win10
Julia 1.6.1
CUDA v3.1.0
```
```
using CUDA
CUDA.device!(0)

function long_gpu_compute(x, num)
   for i = 1:1000000000 x[i % 2 + 1] = num end
   return
end

function f() 
    @cuda threads=1 long_gpu_compute(x, 42.0f0)
    return x 
end 

x = cu(rand(Float32, 2))
for i = 1:3
   println("$i **************************")
   @time( x_out = CUDA.@sync(f(x)))
end
```

If I read documentation correctly @sync meant to be CPU friendly, i.e. equivalent of `CuArrays.@sync` from the past. and the "same" code but using  `CuArrays` i.e `CuArrays.@sync` for syncing is very gentle on CPU. (julia 1.3.1)

For reference the following code is more gentle on CPU (load drops down to practically nothing while waiting for GPU to return).
Also, not sure if that is a practical difference or just an "artificial benchmark effect" below code also runs much faster. 
on 2080Ti
above code 24sec/loop
below code 14sec/loop
```
using CUDAdrv: CuDevice, CuContext, synchronize
use_dev = 0
dev = CuDevice(use_dev)
ctx = CuContext(dev)
println("Running on ", dev)
using CUDAnative
CUDAnative.device!(use_dev)
using CuArrays

function long_gpu_compute(x, num)
    for i = 1:1000000000 x[i % 2 + 1] = num end
   return
end

function f(x) 
    @cuda threads=1 long_gpu_compute(x, 42.0f0)
    return x
end 

x = CuArray{Float32, 1}(undef, 2)

for i = 1:3
   println("$i **************************")
   @time( x_out = CuArrays.@sync(f(x)))
end
```

### PR DESCRIPTION
**Pull request by maleadt**
Fixes https://github.com/JuliaGPU/CUDA.jl/issues/893, fixes https://github.com/JuliaGPU/CUDA.jl/issues/838 and fixes https://github.com/JuliaGPU/CUDA.jl/issues/839.

### CODE REVIEW

**Comment by maleadt** [comment]
About the other issues linked here, there's now a fast path that bails out if the stream is idle already, but further passing `blocking=false` to `synchronize` avoids all blocking when latency is important:

```
julia> @btime begin abs!($y1, $x1); synchronize() end
  7.095 μs (14 allocations: 320 bytes)

julia> @btime begin abs!($y1, $x1); synchronize(blocking=false) end
  4.703 μs (28 allocations: 553 bytes)
```

This is done by benchmarking routines like `CUDA.@time`, where low latency is important for accurate measurements. But in the general case, a couple of us to avoid busy waiting seems like a good trade-off.

**Comment by maleadt** [comment]
I extended the fast path, so that it covers the cases from the linked PRs. So for these, it shouldn't be necessary to pass `blocking=false` to get minimal latency.

File: lib/cudadrv/stream.jl [hunk 1]
```diff
@@ -107,23 +107,39 @@ end
...
+    # minimize latency of short operations by busy-waiting,
+    # initially without even yielding to other tasks
+    spins = 0
+    while blocking || spins < 256
+        if spins < 32
             ccall(:jl_cpu_pause, Cvoid, ())
             # Temporary solution before we have gc transition support in codegen.
             ccall(:jl_gc_safepoint, Cvoid, ())
+            spins += 1
+        else
```
Comment (vchuravy) [inline lib/cudadrv/stream.jl:127]: There is no increment on `spins` in the else path so `spins` won't go beyond 256 and exit the loop.

**Review by vchuravy — CHANGES_REQUESTED** [review]

File: lib/cudadrv/stream.jl (hunk 1 above)
Comment (maleadt) [reply lib/cudadrv/stream.jl:127]: oops

**Review by vchuravy — APPROVED** [review]

**Pull request merged**